### ISSUE TITLE
how long does it take to train mT5 with mC4?

### ISSUE THREAD

**Issue by chattaillaan** [google-research/multilingual-t5#115]
I was given a week to reproduce Google's gT5 with the threat to terminate my PhD study right if I fail, meaning train mT5 with mC4 plus fine-tuning. Isn't that an unreasonable request? The model and dataset are huge.

**Comment by craffel** [comment]
I think by gT5 you mean mT5. Training time will depend on model size and the number of accelerators you use. Pre-training mT5-Base would cost tens of thousands of dollars in compute, but could likely be done in a week. If you have an advisor who threatens to terminate your PhD if you don't complete certain tasks, you should seek a different advisor.

**Comment by chattaillaan** [comment]
thank you, you think that even training with mC4 (almost 40 TB) could be done in a week? 
According to this discussion by the original T5 article writers, downloading and preprocessing the C4 dataset that is 125x smaller than mC4 takes a day. So mC4 would take months.

The [C4][c4] dataset we created for unsupervised pre-training is available in TensorFlow Datasets, but it requires a significant amount of bandwith for downloading the raw [Common Crawl][cc] scrapes (~7 TB) and compute for its preparation (~341 CPU-days). We suggest you take advantage of the [Apache Beam][beam] support in TFDS, which enables distributed preprocessing of the dataset and can be run on [Google Cloud Dataflow][gcd]. With 450 workers, the job should complete in ~18 hours.
https://github.com/tensorflow/datasets/issues/1931

You'll still need to preprocess that 1 cc dump (for one language) which is very large. It takes ~29 hours with 250 DataFlow workers.
https://github.com/tensorflow/datasets/issues/2762

**Comment by craffel** [comment]
C4 and mC4 are both available pre-generated. https://huggingface.co/datasets/mc4

**Comment by chattaillaan** [comment]
But the problem is one cannot use them as such to produce gT5. One needs to heavily preprosess mC4 dataset before training mT5 with it, please see the article about gT5 method/model, 2.2 GEC Pre-training. I checked your link and according to quick browsing it does seem the data is paragraphs not sentences.

The span-prediction objective of mT5 does not enable the model to perform GEC without further fine-tuning, as the span-prediction task uses special tokens to indicate where text should be inserted. Another limiting constraint is that mT5
has been trained on paragraphs, not sentences. We therefore split all paragraphs in mC4 corpus into sentences. We corrupt each sentence using a combination of the following operations:

I was actually given a task to USE gT5 which is not published model so I would need to create it from scratch which according to best of my knowledge took even Google 6-12 months. 
https://aclanthology.org/2021.acl-short.89/

Here a comment from someone trying to use just one language partially:

This is great news! And thanks for making it available. C4 is arguably an expensive and difficult dataset to get, but even more so the multilingual part. I work at the National Library of Norway where I have been trying to download and process mC4 to extract data to train Scandinavian language models (Danish, Icelandic, Norwegian, and Swedish), but just processing the smallest dump was kinda expensive (~950€ using Google Dataflow and Tensorflow Datasets), so processing the ~72 dumps is just not an option. So if y'all have extracted that data and there is a way to access it or transfer it that would be awesome :)

**Comment by craffel** [comment]
Oh, I didn't know what gT5 is, I have no idea about that model. This repository is for mT5, not gT5.

**Comment by chattaillaan** [comment]
Yes, gT5 is created by training mT5 with preprocessed mC4 for GEC + optional fine-tuning. 
Thank you for your insights. 
Also, if someone has something to add, please feel free, it is not so easy to find info about resource requirements.
